// @langchain/ollama/dist/utils/stream_events.cjs
let _langchain_core_language_models_compat = require("@langchain/core/language_models/compat");
//#region src/utils/stream_events.ts
/**
* Converts Ollama chat stream chunks into LangChain ChatModelStreamEvents.
*
* @module
*/
async function* convertOllamaStream(source, options = {}) {
	const shouldStreamUsage = options.streamUsage ?? true;
	const preferThinking = options.think ?? false;
	const blockAccumulators = /* @__PURE__ */ new Map();
	const blockKeyToIndex = /* @__PURE__ */ new Map();
	let nextBlockIndex = 0;
	let messageStarted = false;
	let usageSnapshot;
	let finishReason;
	const getOrCreateBlockIndex = (key, initial) => {
		const existing = blockKeyToIndex.get(key);
		if (existing !== void 0) return {
			index: existing,
			isNew: false
		};
		const index = nextBlockIndex++;
		blockKeyToIndex.set(key, index);
		blockAccumulators.set(index, { ...initial });
		return {
			index,
			isNew: true
		};
	};
	for await (const chunk of source) {
		if (!messageStarted) {
			messageStarted = true;
			yield { event: "message-start" };
		}
		if (shouldStreamUsage) {
			const input = chunk.prompt_eval_count ?? 0;
			const output = chunk.eval_count ?? 0;
			if (input > 0 || output > 0) {
				usageSnapshot = {
					input_tokens: input,
					output_tokens: output,
					total_tokens: input + output
				};
				yield {
					event: "usage",
					usage: usageSnapshot
				};
			}
		}
		if (chunk.done_reason) finishReason = mapOllamaDoneReason(chunk.done_reason);
		const { message } = chunk;
		if (preferThinking && message.thinking) {
			const { index, isNew } = getOrCreateBlockIndex("reasoning", {
				type: "reasoning",
				reasoning: ""
			});
			if (isNew) yield {
				event: "content-block-start",
				index,
				content: {
					type: "reasoning",
					reasoning: ""
				}
			};
			const acc = blockAccumulators.get(index);
			acc.reasoning = (acc.reasoning ?? "") + message.thinking;
			yield {
				event: "content-block-delta",
				index,
				delta: {
					type: "reasoning-delta",
					reasoning: message.thinking
				}
			};
		}
		if (message.content) {
			const { index, isNew } = getOrCreateBlockIndex("text", {
				type: "text",
				text: ""
			});
			if (isNew) yield {
				event: "content-block-start",
				index,
				content: {
					type: "text",
					text: ""
				}
			};
			const acc = blockAccumulators.get(index);
			acc.text = (acc.text ?? "") + message.content;
			yield {
				event: "content-block-delta",
				index,
				delta: {
					type: "text-delta",
					text: message.content
				}
			};
		}
		if (message.tool_calls?.length) for (let i = 0; i < message.tool_calls.length; i++) {
			const tc = message.tool_calls[i];
			const key = `tool:${i}`;
			const args = typeof tc.function.arguments === "string" ? tc.function.arguments : JSON.stringify(tc.function.arguments);
			const { index, isNew } = getOrCreateBlockIndex(key, {
				type: "tool_call_chunk",
				name: tc.function.name,
				args: "",
				index: i
			});
			if (isNew) yield {
				event: "content-block-start",
				index,
				content: {
					type: "tool_call_chunk",
					name: tc.function.name,
					args: "",
					index: i
				}
			};
			const acc = blockAccumulators.get(index);
			acc.name = tc.function.name;
			acc.args = args;
			yield {
				event: "content-block-delta",
				index,
				delta: {
					type: "block-delta",
					fields: {
						type: "tool_call_chunk",
						name: acc.name,
						args: acc.args
					}
				}
			};
		}
	}
	for (const [index, acc] of blockAccumulators) yield {
		event: "content-block-finish",
		index,
		content: (0, _langchain_core_language_models_compat.finalizeContentBlock)(acc)
	};
	yield {
		event: "message-finish",
		reason: finishReason,
		...usageSnapshot ? { usage: usageSnapshot } : {},
		responseMetadata: { model_provider: "ollama" }
	};
}
function mapOllamaDoneReason(reason) {
	switch (reason) {
		case "stop": return "stop";
		case "length": return "length";
		default: return "stop";
	}
}
//#endregion
exports.convertOllamaStream = convertOllamaStream;
